agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+package com.luxoft.blockchainlab.corda.hyperledger.indy
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.node.ObjectNode
+import com.luxoft.blockchainlab.hyperledger.indy.*
+import com.luxoft.blockchainlab.hyperledger.indy.utils.SerializationUtils
+import mu.KotlinLogging
+import org.java_websocket.client.WebSocketClient
+import org.java_websocket.handshake.ServerHandshake
+import java.lang.Thread.sleep
+import java.net.URI
+import java.time.Instant
+import java.util.*
+import kotlin.RuntimeException
+
+
+class PythonRefAgentConnection : AgentConnection {
+    override fun disconnect() {
+        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            webSocket.closeBlocking()
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+        }
+    }
+
+    override fun connect(url: String, login: String, password: String) {
+        disconnect()
+        webSocket = AgentWebSocketClient(URI(url))
+        webSocket.apply {
+            connectBlocking()
+            sendJson(StateRequest())
+            var stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            if(!checkState(stateResponse, login)) {
+                sendJson(WalletConnect(login, password, id = Instant.now().toEpochMilli().toString()))
+                stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            }
+            if(!checkState(stateResponse, login))
```

Comment:
As u can see, we query state 2 times, and check it twice. 
If response returns true, it will mean that wallet is already oppened with for our user.

Reply:
But logic is related to line 34 and not 31? I assume there is should be error handling if 31 return "true"

Reply:
This is by design. Pythonic indy-agent has 2 states: 'initialized' and 'not initialized'. Agent sends it's state within a STATE message, in response to the STATE_REQUEST. If the indy-agent has 'initialized' state, the driver _must not_ send the CONNECT message (along with the login/password) because it causes an error. Otherwise, the agent awaits for a CONNECT message, and replies with STATE which is checked again in the line 36. The reason behind this code is that the client must not send the CONNECT prior to checking the state. If the state is 'not initialized', the client sends a CONNECT and checks the STATE received in reply. This is why the state is checked 2 times.

Reply:
Correct. I see now it can be inside the bracket.